controller/backing_image_controller.go

```diff
@@ -0,0 +1,621 @@
+		} else {
+			switch downloadPod.Status.Phase {
+			case corev1.PodRunning:
+				isReady := false
+				for _, cond := range downloadPod.Status.Conditions {
+					if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+						isReady = true
+					}
+				}
+				if isReady {
+					if backingImage.Status.DiskDownloadStateMap[diskUUID] != types.BackingImageDownloadStateDownloaded {
+						log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Infof("The download state becomes %v from %v since the download pod is ready",
+							types.BackingImageDownloadStateDownloaded, backingImage.Status.DiskDownloadStateMap[diskUUID])
+						backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateDownloaded
+					}
+				} else {
+					if backingImage.Status.DiskDownloadStateMap[diskUUID] == types.BackingImageDownloadStateDownloaded {
+						log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Warnf("The download state becomes %v from %v since the download pod suddenly becomes unavailable",
+							types.BackingImageDownloadStateFailed, types.BackingImageDownloadStateDownloaded)
+						backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+					}
+				}
+			case corev1.PodPending:
+				backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateDownloading
+			default:
+				if backingImage.Status.DiskDownloadStateMap[diskUUID] != types.BackingImageDownloadStateFailed {
+					log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Warnf("The download state becomes %v from %v due to the download pod phase %v",
+						types.BackingImageDownloadStateFailed, backingImage.Status.DiskDownloadStateMap[diskUUID], downloadPod.Status.Phase)
+					backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+				}
+			}
+		}
+
+		// Clean up the failed pod first then recreate it later
+		if backingImage.Status.DiskDownloadStateMap[diskUUID] == types.BackingImageDownloadStateFailed {
+			if downloadPod != nil && downloadPod.DeletionTimestamp == nil {
+				log.WithFields(logrus.Fields{"diskUUID": diskUUID, "pod": downloadPodName}).Infof("Deleting the pod since the downloading state is %v", types.BackingImageDownloadStateFailed)
+				if err := bic.ds.DeletePod(downloadPodName); err != nil {
+					return err
+				}
+			}
+		}
+
+		// Create/Recreate the download pod for a disk if the old pod is gone
+		// and the disk on the node are ready
+		if downloadPod != nil {
+			continue
+		}
+		node, _, err := bic.ds.GetReadyDiskNode(diskUUID)
+		if err != nil {
+			if types.ErrorIsNotFound(err) {
+				backingImage.Status.DiskDownloadStateMap[diskUUID] = types.BackingImageDownloadStateFailed
+				continue
+			}
+			return err
+		}
+		if err := bic.createDownloadPod(downloadPodName, diskUUID, node, backingImage); err != nil {
```

Comment:
Move the node retrieval into the createDownloadPod function and only pass a diskUUID to the function
Then you can pass through the notFound error from there and handle it below as part of the error handling for the createDownloadPod below